wcfsetup/install/files/lib/system/label/LabelPicker.class.php

```diff
@@ -0,0 +1,141 @@
+<?php
+
+namespace wcf\system\label;
+
+use wcf\data\label\group\ViewableLabelGroup;
+use wcf\data\label\Label;
+use wcf\util\JSON;
+use wcf\util\StringUtil;
+
+/**
+ * Provides helper methods to interact with the label group.
+ *
+ * @author Alexander Ebert
+ * @copyright 2001-2023 WoltLab GmbH
+ * @license GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @since 6.1
+ */
+final class LabelPicker
+{
+    /**
+     * Controls the availability of the inverted selection, allowing to filter
+     * for items that are not assigned any label of a label group.
+     */
+    public readonly bool $invertible;
+
+    /**
+     * Name of the hidden input field.
+     */
+    public string $name = 'labelIDs';
```

Comment:
It's probably not possible to make this class entirely readonly, due to `$selected`, but it feels odd that `$name` is mutable, but `$invertible` is not.

PHP does not yet support “clone-with”, which makes the whole ordeal a little more unwieldy, but you might consider adding:

```
public function withName(string $name)
{
  $copy = clone $this;
  $copy->name = $name;

  return $copy;
}
```

and expect users not to touch `$name` directly.

Reply:
This is a short-coming of the existing implementation and the new label picker is intended to be slotted right in with minimal friction. Unfortunately because of how the form builder works, the actual name is not known ahead of time so it must be mutable during its runtime.